File: packages/shield/src/columns.js
/**
 * @module columns
 * Column type detection heuristics for UK school data exports.
 * Pure functions, zero dependencies, ES module.
 */

/**
 * Headers that indicate an always-sensitive identifier column.
 * Matched case-insensitively against the full header string.
 * @type {Array<{pattern: RegExp, label: string}>}
 */
const SENSITIVE_PATTERNS = [
  { pattern: /\bupn\b/i, label: 'UPN' },
  { pattern: /\buln\b/i, label: 'ULN' },
  { pattern: /\bdob\b/i, label: 'DOB' },
  { pattern: /\bdate\s*of\s*birth\b/i, label: 'DOB' },
  { pattern: /\bpostcode\b/i, label: 'postcode' },
  { pattern: /\bemail\b/i, label: 'email' },
  { pattern: /\bphone\b/i, label: 'phone' },
  { pattern: /\bnhs\b/i, label: 'NHS' },
  { pattern: /\baddress\b/i, label: 'address' },
  { pattern: /\bcandidate\s*no\b/i, label: 'candidate number' },
  { pattern: /\badmission\s*no\b/i, label: 'admission number' },
];

/**
 * UK MIS column header patterns that boost categorical classification.
 * @type {RegExp[]}
 */
const MIS_CATEGORICAL_PATTERNS = [
  /\bform\b/i,
  /\bhouse\b/i,
  /\bset\b/i,
  /\bks2\b/i,
  /\bfft\b/i,
  /\bcat\b/i,
  /\bpp\b/i,
  /\bsen\b/i,
  /\beal\b/i,
  /\blac\b/i,
  /\battendance\b/i,
  /\bteacher\b/i,
  /\btutor\b/i,
  /\byear\b/i,
  /\bclass\b/i,
];

/**
 * Confidence thresholds per type. Below these, column is forced to review.
 */
const THRESHOLDS = {
  name: 0.90,
  'free-text': 0.85,
  date: 0.80,
  numerical: 0.75,
  categorical: 0.75,
};

/**
 * Non-numeric exception values that should not disqualify a column as numerical.
 */
const NUMERIC_EXCEPTIONS = new Set(['absent', 'n/a', 'x', '-', '']);

/**
 * Header keywords that boost specific type classifications.
 */
const HEADER_BOOSTS = {
  name: [/\bname\b/i, /\bstudent\b/i, /\bpupil\b/i],
  numerical: [/\bscore\b/i, /\bmark\b/i, /\bpercent\b/i, /\battendance\b/i, /\bgrade\b/i],
  'free-text': [/\bcomment\b/i, /\bnote\b/i, /\bremark\b/i],
  date: [/\bdate\b/i, /\bdob\b/i, /\bbirth\b/i],
};

/** Boost amount added to confidence when header matches. */
const HEADER_BOOST_AMOUNT = 0.10;

/**
 * Check whether a header matches any always-sensitive identifier pattern.
 *
 * @param {string} header
 * @returns {{ sensitive: boolean, label: string|null }}
 */
function checkSensitive(header) {
  for (const { pattern, label } of SENSITIVE_PATTERNS) {
    if (pattern.test(header)) {
      return { sensitive: true, label };
    }
  }
  return { sensitive: false, label: null };
}

/**
 * Check whether a header matches any UK MIS categorical pattern.
 *
 * @param {string} header
 * @returns {boolean}
 */
function isMisCategorical(header) {
  return MIS_CATEGORICAL_PATTERNS.some((p) => p.test(header));
}

/**
 * Check whether a header matches boosts for a given type.
 *
 * @param {string} header
 * @param {string} type
 * @returns {boolean}
 */
function hasHeaderBoost(header, type) {
  const patterns = HEADER_BOOSTS[type];
  if (!patterns) return false;
  return patterns.some((p) => p.test(header));
}

/**
 * Test whether a value looks like a name.
 * Requires 2-4 words, primarily alphabetic characters (plus hyphens, apostrophes, commas, periods, spaces).
 * Also handles surname-first format: "Barker, Dominic".
 * Rejects long strings (sentences/free-text) and code-like short formats.
 *
 * @param {string} val
 * @returns {boolean}
 */
function looksLikeName(val) {
  const trimmed = val.trim();
  if (trimmed.length < 3) return false;

  // Names are short — reject anything that looks like a sentence
  if (trimmed.length > 50) return false;

  // Strip commas for word counting (handles "Barker, Dominic")
  const normalised = trimmed.replace(/,/g, '');

  // Must have 2-4 words (names rarely have more than 4 parts)
  const words = normalised.trim().split(/\s+/);
  if (words.length < 2 || words.length > 4) return false;

  // Each word should be primarily alphabetic (allow hyphens, apostrophes within)
  for (const word of words) {
    // Remove leading/trailing punctuation like periods
    const cleaned = word.replace(/^[.\-']+|[.\-']+$/g, '');
    if (cleaned.length < 1) return false;
    // Must be mostly alpha (allow internal hyphens and apostrophes)
    if (!/^[A-Za-z][A-Za-z'\-]*[A-Za-z]$/.test(cleaned) && !/^[A-Za-z]$/.test(cleaned)) {
      return false;
    }
  }

  // Reject code-like formats: all words too short (e.g. "D.B.", "DBa")
  const allShort = words.every((w) => w.replace(/[^A-Za-z]/g, '').length <= 2);
  if (allShort) return false;

  return true;
}

/**
 * Test whether a value looks like a number (or an acceptable exception).
 *
 * @param {string} val
 * @returns {{ isNumeric: boolean, isException: boolean }}
 */
function looksLikeNumber(val) {
  const trimmed = val.trim();
  if (trimmed === '') return { isNumeric: false, isException: true };

  if (NUMERIC_EXCEPTIONS.has(trimmed.toLowerCase())) {
    return { isNumeric: false, isException: true };
  }

  // Try parsing as number
  const num = Number(trimmed);
  if (!Number.isNaN(num) && trimmed !== '') {
    return { isNumeric: true, isException: false };
  }

  // Allow percentage strings like "85%"
  if (/^\d+(\.\d+)?%$/.test(trimmed)) {
    return { isNumeric: true, isException: false };
  }

  return { isNumeric: false, isException: false };
}

/**
 * Test whether a value matches a date pattern.
 *
 * @param {string} val
 * @returns {boolean}
 */
function looksLikeDate(val) {
  const trimmed = val.trim();
  // dd/mm/yyyy, dd-mm-yyyy, dd.mm.yyyy
  if (/^\d{1,2}[\/\-\.]\d{1,2}[\/\-\.]\d{2,4}$/.test(trimmed)) return true;
  // yyyy-mm-dd
  if (/^\d{4}[\/\-\.]\d{1,2}[\/\-\.]\d{1,2}$/.test(trimmed)) return true;
  // Named months: "15 Jan 2024", "January 15, 2024", "15-Jan-2024"
  if (/\b\d{1,2}[\s\-][A-Za-z]{3,9}[\s\-,]\s*\d{2,4}\b/.test(trimmed)) return true;
  if (/\b[A-Za-z]{3,9}\s+\d{1,2},?\s+\d{2,4}\b/.test(trimmed)) return true;
  return false;
}

/**
 * Classify a single column by analysing header and cell values.
 *
 * @param {string} header - Column header
 * @param {string[]} values - All non-empty cell values for this column
 * @param {number} totalRows - Total number of data rows
 * @returns {{ type: string, confidence: number }}
 */
function classifyColumn(header, values, totalRows) {
  if (values.length === 0) {
    return { type: 'categorical', confidence: 0 };
  }

  const nonEmpty = values.filter((v) => v.trim() !== '');
  if (nonEmpty.length === 0) {
    return { type: 'categorical', confidence: 0 };
  }

  const count = nonEmpty.length;

  // --- Score each type ---

  // Name score
  let nameMatches = 0;
  for (const v of nonEmpty) {
    if (looksLikeName(v)) nameMatches++;
  }
  let nameConf = nameMatches / count;
  if (hasHeaderBoost(header, 'name')) {
    nameConf = Math.min(1, nameConf + HEADER_BOOST_AMOUNT);
  }

  // Date score
  let dateMatches = 0;
  for (const v of nonEmpty) {
    if (looksLikeDate(v)) dateMatches++;
  }
  let dateConf = dateMatches / count;
  if (hasHeaderBoost(header, 'date')) {
    dateConf = Math.min(1, dateConf + HEADER_BOOST_AMOUNT);
  }

  // Numerical score
  let numericMatches = 0;
  let numericExceptions = 0;
  for (const v of nonEmpty) {
    const { isNumeric, isException } = looksLikeNumber(v);
    if (isNumeric) numericMatches++;
    else if (isException) numericExceptions++;
  }
  // Confidence = (numeric + exceptions) / total, but only if there are actual numbers
  let numConf = numericMatches > 0 ? (numericMatches + numericExceptions) / count : 0;
  if (hasHeaderBoost(header, 'numerical')) {
    numConf = Math.min(1, numConf + HEADER_BOOST_AMOUNT);
  }

  // Categorical score
  const uniqueValues = new Set(nonEmpty.map((v) => v.trim().toLowerCase()));
  const uniqueRatio = uniqueValues.size / count;
  const avgLen = nonEmpty.reduce((sum, v) => sum + v.trim().length, 0) / count;

  let catConf = 0;
  if (uniqueRatio < 0.3 && avgLen < 20) {
    catConf = 1 - uniqueRatio; // Fewer unique values = higher confidence
  } else if (uniqueRatio < 0.5 && avgLen < 20) {
    catConf = 0.5;
  }
  if (isMisCategorical(header)) {
    catConf = Math.min(1, catConf + HEADER_BOOST_AMOUNT + 0.15); // Stronger boost for MIS headers
  }

  // Free-text score
  let ftConf = 0;
  if (avgLen > 30 && uniqueRatio > 0.7) {
    ftConf = Math.min(1, uniqueRatio * (avgLen > 50 ? 1 : 0.9));
  } else if (avgLen > 20 && uniqueRatio > 0.5) {
    ftConf = 0.5;
  }
  if (hasHeaderBoost(header, 'free-text')) {
    ftConf = Math.min(1, ftConf + HEADER_BOOST_AMOUNT);
  }

  // --- Priority-based selection: name > date > numerical > categorical > free-text ---
  const candidates = [
    { type: 'name', confidence: nameConf },
    { type: 'date', confidence: dateConf },
    { type: 'numerical', confidence: numConf },
    { type: 'categorical', confidence: catConf },
    { type: 'free-text', confidence: ftConf },
  ];

  // Filter to candidates that meet their threshold
  const passing = candidates.filter(
    (c) => c.confidence >= (THRESHOLDS[c.type] || 0.75)
  );

  if (passing.length > 0) {
    // Return the first passing candidate (priority order is already correct)
    return { type: passing[0].type, confidence: passing[0].confidence };
  }

  // Nothing passes threshold — pick the best confidence and flag for review
  candidates.sort((a, b) => b.confidence - a.confidence);
  const best = candidates[0];
  // If everything is 0, default to categorical
  if (best.confidence === 0) {
    return { type: 'categorical', confidence: 0 };
  }
  return { type: best.type, confidence: best.confidence };
}

/**
 * Detect column types for a tabular dataset.
 *
 * @param {string[]} headers - Array of column header strings.
 * @param {Array<Array<string>>} rows - Array of row arrays (strings), no header row.
 * @returns {{ columns: Array<{index: number, name: string, type: string, confidence: number, sensitive: boolean, reviewRequired: boolean}> }}
 */
export function detectColumns(headers, rows) {
  if (!headers || headers.length === 0) {
    return { columns: [] };
  }

  const totalRows = rows ? rows.length : 0;
  const columns = [];

  for (let i = 0; i < headers.length; i++) {
    const header = headers[i] || '';
    const values = rows ? rows.map((row) => String(row[i] ?? '')) : [];

    // Check always-sensitive identifiers first
    const { sensitive } = checkSensitive(header);

    // Classify the column content
    const { type: detectedType, confidence } = classifyColumn(header, values, totalRows);

    // If the column is a sensitive identifier, override type
    const type = sensitive ? 'identifier' : detectedType;

    // Determine review requirement based on asymmetric thresholds
    const threshold = THRESHOLDS[type] || 0.75;
    const reviewRequired = !sensitive && confidence < threshold;

    columns.push({
      index: i,
      name: header,
      type,
      confidence: Math.floor(confidence * 100 + 0.5) / 100,
      sensitive,
      reviewRequired,
    });
  }

  return { columns };
}

File: packages/shield-py/shield/columns.py
"""
Column type detection heuristics for UK school data exports.
Pure functions, stdlib only.
"""

import math
import re
from typing import Any, Dict, List, Optional, Tuple


# Headers that indicate an always-sensitive identifier column
_SENSITIVE_PATTERNS = [
    (re.compile(r'\bupn\b', re.IGNORECASE), 'UPN'),
    (re.compile(r'\buln\b', re.IGNORECASE), 'ULN'),
    (re.compile(r'\bdob\b', re.IGNORECASE), 'DOB'),
    (re.compile(r'\bdate\s*of\s*birth\b', re.IGNORECASE), 'DOB'),
    (re.compile(r'\bpostcode\b', re.IGNORECASE), 'postcode'),
    (re.compile(r'\bemail\b', re.IGNORECASE), 'email'),
    (re.compile(r'\bphone\b', re.IGNORECASE), 'phone'),
    (re.compile(r'\bnhs\b', re.IGNORECASE), 'NHS'),
    (re.compile(r'\baddress\b', re.IGNORECASE), 'address'),
    (re.compile(r'\bcandidate\s*no\b', re.IGNORECASE), 'candidate number'),
    (re.compile(r'\badmission\s*no\b', re.IGNORECASE), 'admission number'),
]

# UK MIS column header patterns that boost categorical classification
_MIS_CATEGORICAL_PATTERNS = [
    re.compile(r'\bform\b', re.IGNORECASE),
    re.compile(r'\bhouse\b', re.IGNORECASE),
    re.compile(r'\bset\b', re.IGNORECASE),
    re.compile(r'\bks2\b', re.IGNORECASE),
    re.compile(r'\bfft\b', re.IGNORECASE),
    re.compile(r'\bcat\b', re.IGNORECASE),
    re.compile(r'\bpp\b', re.IGNORECASE),
    re.compile(r'\bsen\b', re.IGNORECASE),
    re.compile(r'\beal\b', re.IGNORECASE),
    re.compile(r'\blac\b', re.IGNORECASE),
    re.compile(r'\battendance\b', re.IGNORECASE),
    re.compile(r'\bteacher\b', re.IGNORECASE),
    re.compile(r'\btutor\b', re.IGNORECASE),
    re.compile(r'\byear\b', re.IGNORECASE),
    re.compile(r'\bclass\b', re.IGNORECASE),
]

# Confidence thresholds per type
_THRESHOLDS = {
    'name': 0.90,
    'free-text': 0.85,
    'date': 0.80,
    'numerical': 0.75,
    'categorical': 0.75,
}

# Non-numeric exception values
_NUMERIC_EXCEPTIONS = {'absent', 'n/a', 'x', '-', ''}

# Header keywords that boost specific type classifications
_HEADER_BOOSTS = {
    'name': [re.compile(r'\bname\b', re.IGNORECASE), re.compile(r'\bstudent\b', re.IGNORECASE), re.compile(r'\bpupil\b', re.IGNORECASE)],
    'numerical': [re.compile(r'\bscore\b', re.IGNORECASE), re.compile(r'\bmark\b', re.IGNORECASE), re.compile(r'\bpercent\b', re.IGNORECASE), re.compile(r'\battendance\b', re.IGNORECASE), re.compile(r'\bgrade\b', re.IGNORECASE)],
    'free-text': [re.compile(r'\bcomment\b', re.IGNORECASE), re.compile(r'\bnote\b', re.IGNORECASE), re.compile(r'\bremark\b', re.IGNORECASE)],
    'date': [re.compile(r'\bdate\b', re.IGNORECASE), re.compile(r'\bdob\b', re.IGNORECASE), re.compile(r'\bbirth\b', re.IGNORECASE)],
}

_HEADER_BOOST_AMOUNT = 0.10


def _check_sensitive(header: str) -> Tuple[bool, Optional[str]]:
    """Check whether a header matches any always-sensitive identifier pattern."""
    for pattern, label in _SENSITIVE_PATTERNS:
        if pattern.search(header):
            return True, label
    return False, None


def _is_mis_categorical(header: str) -> bool:
    """Check whether a header matches any UK MIS categorical pattern."""
    return any(p.search(header) for p in _MIS_CATEGORICAL_PATTERNS)


def _has_header_boost(header: str, col_type: str) -> bool:
    """Check whether a header matches boosts for a given type."""
    patterns = _HEADER_BOOSTS.get(col_type)
    if not patterns:
        return False
    return any(p.search(header) for p in patterns)


def _looks_like_name(val: str) -> bool:
    """Test whether a value looks like a name."""
    trimmed = val.strip()
    if len(trimmed) < 3:
        return False
    if len(trimmed) > 50:
        return False

    # Strip commas for word counting
    normalised = trimmed.replace(',', '')
    words = normalised.strip().split()
    if len(words) < 2 or len(words) > 4:
        return False

    for word in words:
        # Remove leading/trailing punctuation
        cleaned = re.sub(r"^[.\-']+|[.\-']+$", '', word)
        if len(cleaned) < 1:
            return False
        if not re.match(r"^[A-Za-z][A-Za-z'\-]*[A-Za-z]$", cleaned) and not re.match(r'^[A-Za-z]$', cleaned):
            return False

    # Reject code-like formats: all words too short
    all_short = all(len(re.sub(r'[^A-Za-z]', '', w)) <= 2 for w in words)
    if all_short:
        return False

    return True


def _looks_like_number(val: str) -> Dict[str, bool]:
    """Test whether a value looks like a number."""
    trimmed = val.strip()
    if trimmed == '':
        return {'isNumeric': False, 'isException': True}

    if trimmed.lower() in _NUMERIC_EXCEPTIONS:
        return {'isNumeric': False, 'isException': True}

    # Try parsing as number
    try:
        num = float(trimmed)
        if math.isfinite(num):
            return {'isNumeric': True, 'isException': False}
    except ValueError:
        pass

    # Allow percentage strings like "85%"
    if re.match(r'^\d+(\.\d+)?%$', trimmed):
        return {'isNumeric': True, 'isException': False}

    return {'isNumeric': False, 'isException': False}


def _looks_like_date(val: str) -> bool:
    """Test whether a value matches a date pattern."""
    trimmed = val.strip()
    if re.match(r'^\d{1,2}[/\-\.]\d{1,2}[/\-\.]\d{2,4}$', trimmed):
        return True
    if re.match(r'^\d{4}[/\-\.]\d{1,2}[/\-\.]\d{1,2}$', trimmed):
        return True
    if re.search(r'\b\d{1,2}[\s\-][A-Za-z]{3,9}[\s\-,]\s*\d{2,4}\b', trimmed):
        return True
    if re.search(r'\b[A-Za-z]{3,9}\s+\d{1,2},?\s+\d{2,4}\b', trimmed):
        return True
    return False


def _classify_column(
    header: str,
    values: List[str],
    total_rows: int,
) -> Dict[str, Any]:
    """Classify a single column by analysing header and cell values."""
    if not values:
        return {'type': 'categorical', 'confidence': 0}

    non_empty = [v for v in values if v.strip() != '']
    if not non_empty:
        return {'type': 'categorical', 'confidence': 0}

    count = len(non_empty)

    # Name score
    name_matches = sum(1 for v in non_empty if _looks_like_name(v))
    name_conf = name_matches / count
    if _has_header_boost(header, 'name'):
        name_conf = min(1, name_conf + _HEADER_BOOST_AMOUNT)

    # Date score
    date_matches = sum(1 for v in non_empty if _looks_like_date(v))
    date_conf = date_matches / count
    if _has_header_boost(header, 'date'):
        date_conf = min(1, date_conf + _HEADER_BOOST_AMOUNT)

    # Numerical score
    numeric_matches = 0
    numeric_exceptions = 0
    for v in non_empty:
        result = _looks_like_number(v)
        if result['isNumeric']:
            numeric_matches += 1
        elif result['isException']:
            numeric_exceptions += 1
    num_conf = (numeric_matches + numeric_exceptions) / count if numeric_matches > 0 else 0
    if _has_header_boost(header, 'numerical'):
        num_conf = min(1, num_conf + _HEADER_BOOST_AMOUNT)

    # Categorical score
    unique_values = set(v.strip().lower() for v in non_empty)
    unique_ratio = len(unique_values) / count
    avg_len = sum(len(v.strip()) for v in non_empty) / count

    cat_conf = 0.0
    if unique_ratio < 0.3 and avg_len < 20:
        cat_conf = 1 - unique_ratio
    elif unique_ratio < 0.5 and avg_len < 20:
        cat_conf = 0.5
    if _is_mis_categorical(header):
        cat_conf = min(1, cat_conf + _HEADER_BOOST_AMOUNT + 0.15)

    # Free-text score
    ft_conf = 0.0
    if avg_len > 30 and unique_ratio > 0.7:
        ft_conf = min(1, unique_ratio * (1 if avg_len > 50 else 0.9))
    elif avg_len > 20 and unique_ratio > 0.5:
        ft_conf = 0.5
    if _has_header_boost(header, 'free-text'):
        ft_conf = min(1, ft_conf + _HEADER_BOOST_AMOUNT)

    # Priority-based selection
    candidates = [
        {'type': 'name', 'confidence': name_conf},
        {'type': 'date', 'confidence': date_conf},
        {'type': 'numerical', 'confidence': num_conf},
        {'type': 'categorical', 'confidence': cat_conf},
        {'type': 'free-text', 'confidence': ft_conf},
    ]

    # Filter to candidates that meet their threshold
    passing = [c for c in candidates if c['confidence'] >= _THRESHOLDS.get(c['type'], 0.75)]

    if passing:
        return {'type': passing[0]['type'], 'confidence': passing[0]['confidence']}

    # Nothing passes — pick best confidence
    candidates.sort(key=lambda c: c['confidence'], reverse=True)
    best = candidates[0]
    if best['confidence'] == 0:
        return {'type': 'categorical', 'confidence': 0}
    return {'type': best['type'], 'confidence': best['confidence']}


def detect_columns(
    headers: List[str],
    rows: List[List[str]],
) -> Dict[str, Any]:
    """Detect column types for a tabular dataset.

    Args:
        headers: Array of column header strings.
        rows: Array of row arrays (strings), no header row.

    Returns:
        Dict with 'columns' list.
    """
    if not headers:
        return {'columns': []}

    total_rows = len(rows) if rows else 0
    columns = []

    for i, header in enumerate(headers):
        header = header or ''
        values = [str(row[i]) if i < len(row) and row[i] is not None else '' for row in (rows or [])]

        # Check always-sensitive identifiers first
        sensitive, label = _check_sensitive(header)

        # Classify the column content
        classification = _classify_column(header, values, total_rows)
        detected_type = classification['type']
        confidence = classification['confidence']

        # If the column is a sensitive identifier, override type
        col_type = 'identifier' if sensitive else detected_type

        # Determine review requirement
        threshold = _THRESHOLDS.get(col_type, 0.75)
        review_required = not sensitive and confidence < threshold

        columns.append({
            'index': i,
            'name': header,
            'type': col_type,
            'confidence': math.floor(confidence * 100 + 0.5) / 100,
            'sensitive': sensitive,
            'reviewRequired': review_required,
        })

    return {'columns': columns}
